### PR TITLE
Standard errors

### PR DESCRIPTION
**Pull request by ialarmedalien**
Standardised error responses from the API.

Also made misc fixes and added tests for various pieces of untested code.

Closes https://github.com/kbase/relation_engine/issues/14

- [x] I updated the README.md docs to reflect this change.
- [x] This is not a breaking API change

### CODE REVIEW

File: relation_engine_server/test/test_spec_loader.py [hunk 1]
```diff
@@ -24,7 +24,8 @@ def test_get_names(self, schema_type_names=[], expected=[]):
 
         # this method should only be run from another test method
         if len(schema_type_names) == 0:
-            self.skipTest('No schema type names supplied. Skipping')
+            self.assertTrue(True)
+            return
```
Comment (ialarmedalien) [inline relation_engine_server/test/test_spec_loader.py:28]: Same effect as `skipTest` but without giving a misleading `s` in the test output

File: scripts/run_tests.sh [hunk 2]
```diff
@@ -13,10 +14,12 @@ sh /app/scripts/start_server.sh &
 # spec validation
 python -m spec.validate &&
 # spec stored query tests
-python -m unittest discover spec/test &&
+coverage run --parallel-mode -m unittest discover spec/test &&
```
Comment (ialarmedalien) [inline scripts/run_tests.sh:17]: Add in coverage calculation

File: spec/test/djornl/results.json [hunk 3]
```diff
@@ -57,34 +57,36 @@
...
-      "As75__AT1G01020__AraGWAS-Phenotype_Associations__39.9",
-      "AT1G01010__AT1G01020__AraNetv2-HT_high-throughput-ppi__2.3",
-      "AT1G01010__AT1G01030__AraNetv2-HT_high-throughput-ppi__2.4",
-      "AT1G01010__AT1G01040__AraNetv2-DC_domain-co-occurrence__2.5",
-      "AT1G01010__AT1G01040__AraNetv2-LC_lit-curated-ppi__170.5",
-      "AT1G01030__AT1G01050__AraNetv2-CX_pairwise-gene-coexpression__2.6",
-      "AT1G01050__AT1G01060__AraNetv2-LC_lit-curated-ppi__2.7",
-      "AT1G01080__AT1G01090__AraNetv2-LC_lit-curated-ppi__2.8"
-    ]
+    "-": {
```
Comment (ialarmedalien) [inline spec/test/djornl/results.json:60]: small reformat, big diff!

File: spec/test/sample_schemas/duplicate_names/ncbi/ncbi_taxon.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+name: ncbi_taxon
```
Comment (ialarmedalien) [inline spec/test/sample_schemas/duplicate_names/ncbi/ncbi_taxon.yaml:1]: Next four files are just to check that the validator picks up duplicate file names.

File: spec/validate.py [hunk 5]
```diff
@@ -36,36 +36,92 @@
...
+                name = data['name']
+                if name in names:
+                    raise ValueError(f"Duplicate queries named '{name}'")
+                else:
+                    names.add(name)
+
+            except Exception as err:
+                print(f"✕ {path} failed validation")
+                print(err)
+                err_count += 1
```
Comment (ialarmedalien) [inline spec/validate.py:72]: validate all files in the dir, don't quit after the first failure

File: spec/validate.py [hunk 6]
```diff
@@ -36,36 +36,92 @@
...
+
+    if err_count:
+        raise ValidationError(f'{directory} failed validation')
+
+    # all's well
+    print('...all valid.')
+    return
+
+
+def validate_all_by_type(validation_base_dir=None):
```
Comment (ialarmedalien) [inline spec/validate.py:82]: new method that pulls together all the validation routines

File: spec/validate.py [hunk 7]
```diff
@@ -184,16 +240,12 @@ def validate_aql_on_arango(data):
...
 
-if __name__ == '__main__':
     wait_for_arangodb()
-    try:
-        for s in ['data_source', 'stored_query', 'view', 'collection']:
-            validate_all(s)
-    except Exception as err:
-        _fatal(err)
+    n_errors = validate_all_by_type(validation_base_dir)
+    sys.exit(n_errors)
```
Comment (jayrbolton) [inline spec/validate.py:251]: Usually each exit code has a special meaning. And it is always limited to one byte, 0-255. So we want to have exit code 1 here.

File: relation_engine_server/main.py [hunk 8]
```diff
@@ -18,6 +18,26 @@
...
 
 
+def return_error(error_dict, code):
+    """return the appropriate error structure and code
+
+    Errors returned by the server have the basic format
+
+    'error': {
+        'message': <text explanation of error>,
+        'status': <HTTP error code>,
```
Comment (jayrbolton) [inline relation_engine_server/main.py:28]: It doesn't seem like we need status here if it is always in the header.

**Comment by jayrbolton** [comment]
A couple small comments above, otherwise looks good

File: relation_engine_server/main.py (hunk 8 above)
Comment (ialarmedalien) [reply relation_engine_server/main.py:28]: Yeah, that was my thought too - the only reason I left it in is that it was in several of the error structures. I'm happy to remove it globally.

File: relation_engine_server/test/test_api_v1.py [hunk 9]
```diff
@@ -96,20 +99,20 @@ def test_request(self, url=None, params=None, data=None, headers=None, method='g
 
     def test_root(self):
         """Test root path for api."""
-        resp = requests.get(URL + '/').json()
-        self.assertEqual(resp['arangodb_status'], 'connected_authorized')
-        self.assertTrue(resp['commit_hash'])
-        self.assertTrue(resp['repo_url'])
+        resp_json = requests.get(URL + '/').json()
```
Comment (ialarmedalien) [inline relation_engine_server/test/test_api_v1.py:102]: Consistent naming for responses -- if it is a response object, call it `resp`; if it is the deJSONified response body, call it `resp_json`.

File: relation_engine_server/test/test_api_v1.py [hunk 10]
```diff
@@ -363,63 +363,83 @@ def test_show_data_source_unknown(self):
...
                 'error': {
-                    'status': 404,
                     'message': 'Not found',
                     'details': f"Data source '{name}' does not exist.",
                     'name': name,
                 }
             }
         )
 
-        resp = requests.get(f"{API_URL}/data_sources/{name}")
```
Comment (ialarmedalien) [inline relation_engine_server/test/test_api_v1.py:374]: repeated test removed

**Pull request merged**